staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -711,28 +809,44 @@ func (c *Cacher) startDispatching(event *watchCacheEvent) {
 	// gain from avoiding memory allocations is much bigger.
 	c.watchersBuffer = c.watchersBuffer[:0]
 
-	// Iterate over "allWatchers" no matter what the trigger function is.
-	for _, watcher := range c.watchers.allWatchers {
-		c.watchersBuffer = append(c.watchersBuffer, watcher)
-	}
-	if supported {
-		// Iterate over watchers interested in the given values of the trigger.
-		for _, triggerValue := range triggerValues {
-			for _, watcher := range c.watchers.valueWatchers[triggerValue] {
-				c.watchersBuffer = append(c.watchersBuffer, watcher)
+	if event.Type == watch.Bookmark {
+		// Pop up watchers those already expired, ignore those already stopped
+		// as we do not delete the watcher when it is stopped.
```

Comment:
nit: maybe
Pop already expired watchers. However, explicitly ignore stopped ones, as we don't delete watcher from bookmarkWatchers when it is stopped.